pkg/utils/termination/termination.go

```diff
@@ -20,62 +20,34 @@ import (
 	"context"
 	"fmt"
 
+	"k8s.io/apimachinery/pkg/api/equality"
+
 	"sigs.k8s.io/controller-runtime/pkg/client"
 
 	v1 "sigs.k8s.io/karpenter/pkg/apis/v1"
 	"sigs.k8s.io/karpenter/pkg/cloudprovider"
 )
 
-// EnsureTerminated is a helper function that takes a v1.NodeClaim and calls cloudProvider.Delete() if status condition
-// on nodeClaim is not terminating. If it is terminating then it will call cloudProvider.Get() to check if the instance
-// is terminated or not. It will return an error and a boolean that indicates if the instance is terminated or not. We simply return
-// conflict or a NotFound error if we encounter it while updating the status on nodeClaim.
+// EnsureTerminated is a helper function that takes a v1.NodeClaim and calls cloudProvider.Delete(). It will return an
+// error and a boolean that indicates if the instance is terminated or not. We simply return conflict or a NotFound
+// error if we encounter it while updating the status on nodeClaim.
 func EnsureTerminated(ctx context.Context, c client.Client, nodeClaim *v1.NodeClaim, cloudProvider cloudprovider.CloudProvider) (terminated bool, err error) {
-	// Check if the status condition on nodeClaim is Terminating
-	if !nodeClaim.StatusConditions().Get(v1.ConditionTypeInstanceTerminating).IsTrue() {
-		// If not then call Delete on cloudProvider to trigger termination and always requeue reconciliation
-		if err = cloudProvider.Delete(ctx, nodeClaim); err != nil {
-			if cloudprovider.IsNodeClaimNotFoundError(err) {
-				stored := nodeClaim.DeepCopy()
-				updateStatusConditionsForDeleting(nodeClaim)
-				// We use client.MergeFromWithOptimisticLock because patching a list with a JSON merge patch
-				// can cause races due to the fact that it fully replaces the list on a change
-				// Here, we are updating the status condition list
-				if err = c.Status().Patch(ctx, nodeClaim, client.MergeFromWithOptions(stored, client.MergeFromWithOptimisticLock{})); err != nil {
-					return false, err
-				}
-				// Instance is terminated
-				return true, nil
+	stored := nodeClaim.DeepCopy()
+	err = cloudProvider.Delete(ctx, nodeClaim)
+	// Set InstanceTerminating to true when cloudProvider.Delete() returns -
+	// 1. nodeClaim not found error which indicates the instance was terminated
+	// 2. no error which indicates that Delete() has been processed but the instance has not terminated so requeue
```

Comment:
i think it would be nice to update the code comment on the `Delete` interface in https://github.com/kubernetes-sigs/karpenter/blob/main/pkg/cloudprovider/types.go#L60 to ensure that provider developers understand the expectation. eg "delete should return nil if the instance is terminating and NotFoundError if the instance is deleted".

Reply:
Right, makes sense. I will update the comment.